### PR TITLE
rework addon-resizer to use client-go 

### PR DESCRIPTION
**Pull request by piontec**
The addon-resizer code doesn't modify attributes other than limits/requests, so it seems the bug is in libraries used. This is causing problems like #1858 and probably also #1799 and #1457 .  To solve that, I upgraded libraries to more recent ones. 

Changes include:
* drop `godep` in favor of `go mod` - I was becoming sick trying to do that using `godep`
* code is reworked to use `client-go` (version v8.0.0) and api libs (from 1.11 k8s release)
* go version is updated to 1.12.1
* `Makefile` builds docker image OK

Testing:
* unit tests are passing
* I manually verified that `addon-scaler` works OK running against an already provisioned cluster with `kube-state-metrics` pod.

### CODE REVIEW

**Comment by piontec** [comment]
Assuming the switch to `go mod` is OK, we can drop `vendor/` in the repo and fetch dependencies on build time. Let me know if that's ok.

**Comment by bskiba** [comment]
In general, huge, huge thank you for taking the time and the effort!

Seeing core Kubernetes moving to go mod (https://groups.google.com/forum/#!topic/kubernetes-dev/ECejidBCELI) I think it's the right thing to move addon resizer too.

I have one request, can you split the go mod change from the client-go one? It will make it way easier for me to remove the code that actually changed.

**Comment by piontec** [comment]
Crap, I squashed my commits - I can't easily separate them now. But if we remove `vendor/` and rely entirely on the new `go mod`, the changes should be much easier to review - the whole switch will be just in removing `Godeps/` and adding `go.mod`. What do you think?

**Comment by bskiba** [comment]
There are several problems:
- I need to make sure the dependencies haven't changed - we are not pulling in problematic licenses
- I believe we need to have a copy of the license for all vendored dependencies

I can try to review this as a whole but I'll need time. I also need to consult someone around the license thing, I'll get back to you next week.
It's easy to omit things with such a huge change. Example: you have two different go versions specified, one in Makefile (1.12) and another in go.mod (1.11). Is that intentional?

**Comment by piontec** [comment]
Well, dependencies have changed, that's for sure. `client-go` wasn't even used before and it is now. The same with `apimachinery`. The problem with this project was that dependencies were so old and the project so small, that I actually removed all dependencies, started with empty `go.mod` and then I was rewriting the code to use new libraries, adding them to `go.mod` on the way.

As for version mismatch - my mistake, I forgot to switch everything to 1.12. Will fix in a minute.

**Comment by piontec** [comment]
OK, I'll try to rework using `godeps`. Do you have some hints about how to include `client-go` with it the easiest way?

**Comment by piontec** [comment]
@bskiba Done, it's using `godep` now. Can you also have a look at https://github.com/kubernetes/autoscaler/pull/1843, it seems to be stuck a little...

File: addon-resizer/nanny/kubernetes_client.go [hunk 1]
```diff
@@ -20,40 +20,59 @@ import (
...
+		deploymentClient: kubeClient.AppsV1().Deployments(namespace),
 	}
-	return uint64(len(k.nodeStore.List())), nil
+	return result
 }
 
-func (k *kubernetesClient) ContainerResources() (*apiv1.ResourceRequirements, error) {
-	pod, err := k.clientset.CoreClient.Pods(k.namespace).Get(k.pod)
+func (k *kubernetesClient) Stop() {
+	for i := 0; i < 3; i++ {
```
Comment (bskiba) [inline addon-resizer/nanny/kubernetes_client.go]: The i < 3 seems really error prone - if someone adds another lister, no way they will know thay have to increase it. Can you add a stopChan per lister instead?

File: addon-resizer/nanny/kubernetes_client.go [hunk 2]
```diff
@@ -20,40 +20,59 @@ import (
...
 
-	api "k8s.io/kubernetes/pkg/api"
-	apiv1 "k8s.io/kubernetes/pkg/api/v1"
-	cache "k8s.io/kubernetes/pkg/client/cache"
-	client "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
-	runtime "k8s.io/kubernetes/pkg/runtime"
-	wait "k8s.io/kubernetes/pkg/util/wait"
-	watch "k8s.io/kubernetes/pkg/watch"
+	apps "k8s.io/api/apps/v1"
+	api "k8s.io/api/core/v1"
```
Comment (bskiba) [inline addon-resizer/nanny/kubernetes_client.go]: nit: s/api/core/

File: addon-resizer/main.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+func buildOutOfClusterConfig() (*rest.Config, error) {
+	kubeconfigPath := os.Getenv("KUBECONFIG")
+	if kubeconfigPath == "" {
+		kubeconfigPath = os.Getenv("HOME") + "/.kube/config"
+	}
+	return clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+}
+
+// GetClientOutOfCluster returns a k8s clientset to the request from outside of cluster
+func GetClientOutOfCluster() kubernetes.Interface {
```
Comment (bskiba) [inline addon-resizer/main.go]: nit: GetClientOutOfClusterOrDie() (here and below)

Comment (piontec) [reply addon-resizer/main.go]: GetClientOutOfCluster() has a fallback to GetClient(). I'm changing the latter one to "OrDie()"

**Comment by piontec** [comment]
@bskiba please check now

**Comment by piontec** [comment]
Hi @bskiba , do you have any time frame for releasing this?

**Comment by diclophis** [comment]
Also where can we find newer tagged images of this addon-resizer docker image?

**Comment by bskiba** [comment]
@piontec I'll try to release in the next two weeks.
@diclophis All images should be in k8s.gcr.io/addon-resizer, the newest is 2.1

**Comment by wojtek-t** [comment]
@bskiba - note that we're still using 1.8.* releases, as there were some issues with 2.* releases
I can't remember the details, but it's something to keep in mind.

**Comment by bskiba** [comment]
@wojtek-t Thanks! I wasn't planning on changing the default, and I think that there are folks successfully using the 2.X releases so it still makes sense to release. But it's good to note that maybe this fix should go to the 1.8 release as well. @piontec are you interested in porting to 1.8 or is the master branch satisfactory?

**Pull request merged**